cheatsheets/gleam-for-go-users.md

```diff
@@ -0,0 +1,1460 @@
+
+Since there is no special syntax to create a module, there can be only one
+module in a file and since there is no way name the module the filename
+always matches the module name which keeps things simple and transparent.
+
+In `/src/foo/bar.gleam`:
+
+```gleam
+// Creation of module function identity
+// in module bar
+pub fn identity(x) {
+  x
+}
+```
+
+Importing the `bar` module and calling a module function:
+
+```gleam
+// In src/main.gleam
+import foo/bar // if foo was in a directory called `lib` the import would be `lib/foo/bar`.
+
+pub fn main() {
+  bar.identity(1) // 1
+}
+```
+
+### Imports
+
+#### Go
+
+Go imports are fairly simple.  Using `import` keyword with either a string literal
+that contains the path to a package, or several such strings each on separate lines
+inside of parentheses.
+
+Besides the standard library, all packages have to be fully qualified, where the
+qualifier is the name of the containing module, which by convention is the url
+at which the module is hosted.
+
+```go
+import "fmt" // Single import.  Stdlib packages are referenced bare, no path required
+import ( // Multi-package import
+  "strings" // Available for use as `strings`
+
+  "github.com/example/example-go-project/somepkg" // Will be available as `somepkg`
+)
+```
+
+There is no way in Go to import a symbol itself, all imports are qualified by their
+package.
+
+```go
+fmt.Println()
+strings.Trim()
+```
+
+#### Gleam
+
+Imports are relative to the app `src` folder.
+
+Modules in the same directory will need to reference the entire path from `src`
+for the target module, even if the target module is in the same folder.
+
+Inside module `src/nasa/moon_base.gleam`:
+
+```gleam
+// imports module src/nasa/rocket_ship.gleam
+import nasa/rocket_ship
+
+pub fn explore_space() {
+  rocket_ship.launch()
+}
+```
+
+### Named imports
+
+#### Go
+
+Go allows you to alias a package import
+
+```go
+import (
+  other "github.com/example/example-go-project/somepkg"
+)
+
+func main() {
+  other.SomeFunc()
+}
+```
+
+#### Gleam
+
+Gleam has as similar feature:
+
+```gleam
+import unix/cat
+import animal/cat as kitty
+// cat and kitty are available
+```
+
+This may be useful to differentiate between multiple modules that would have the same default name when imported.
+
+### Unqualified imports
+
+#### Go
+
+Go does not support unqualified imports
+
+#### Gleam
+
+```gleam
+import animal/cat.{
+  Cat,
+  stroke
+}
+
+pub fn main() {
+  let kitty = Cat(name: "Nubi")
+  stroke(kitty)
+}
+```
+
+Importing common types such as `gleam/order.{Lt, Eq, Gt}` or
+`gleam/option.{Some,None}` can be very helpful.
+
+## Architecture
+
+To iterate a few foundational differences:
+
+1. Programming model: imperative VS functional immutable
+   programming
+2. Guarantees: implicit nillability vs explicit nillability
+3. Runtime model: lightweight coroutines(`goroutines`) VS Erlang/OTP processes
+4. Error handling: error value (and very few panics) VS result type
```

Comment:
I think Go and Gleam use panics in a similar way- for fatal unexpected errors. I could be wrong though, my Go is rusty.

Reply:
In general, panics should never be used in Go (especially in libraries), you'd return wrapped errors to build a manual one-line stack trace and handle it at the top.

Only time I've used panic is because I was developing quickly and planned to replace it all with error returns

Reply:
That's true, it's rare that you will write a panic, panics will occur, such as when you read from a nil channel

Reply:
Actually I think @lpil is more or less right, a panic can be recovered and is a way for the application to choose to recover and get back into a working state.  Which is similar to having a process being killed in a supervision tree.